### ISSUE TITLE
Verification Regressions with Dafny 4.0

### ISSUE THREAD

**Issue by DavePearce** [dafny-lang/dafny#3758]
### Dafny version

4.0.0

### Code to produce this issue

```dafny
const G_MEMORY: nat := 3

function QuadraticCost(memUsedSize: nat): nat {
  G_MEMORY * memUsedSize + ((memUsedSize * memUsedSize) / 512)
}


/**
 *  The quadratic cost function is increasing.
 */
lemma QuadraticCostIsMonotonic_1(x: nat, y: nat)
ensures x >= y ==> QuadraticCost(x) >= QuadraticCost(y)
{
    if x >= y {
        calc >= {
            G_MEMORY * x + ((x * x) / 512);
            G_MEMORY * y + ((y * y) / 512);
        }
    }
}

/**
 *  The quadratic cost function is increasing.
 */
lemma QuadraticCostIsMonotonic_2(x: nat, y: nat)
  ensures x >= y ==> QuadraticCost(x) >= QuadraticCost(y)
{
  if x > y {
    QuadraticCostIsMonotonic_2(x-1,y);
  }
}
```

### What happened?

Since upgrading to Dafny 4.0.0 we are experiencing a wide range of verification failures on our code base (approx 8KLOC including comments and whitespace), and I wanted to understand whether others are having similar issues or whether there are any known solutions.

The example given above is illustrates the kind of issues we've experienced.  The first version `QuadraticCostIsMonotonic_1()` is the lemma we had in our code base that previously verified with Dafny `3.13.0`.  The second version (`QuadraticCostIsMonotonic_2()`) is a version of the lemma which now verifies with Dafny `4.0.0`.  Both versions verify under Dafny `3.13.0`.  Some interesting observations though:

- Under Dafny `4.0.0` **with** the switch `/arith:2` then `QuadracticCostIsMonotonic_1()` verifies, but  `QuadracticCostIsMonotonic_2()` does not.
- Under Dafny `4.0.0` **without** the switch `/arith:2` then `QuadracticCostIsMonotonic_2()` verifies, but  `QuadracticCostIsMonotonic_1()` does not.

I should further note that, with the switch `/arith:9`, then neither will verify.  Likewise, `/noNLarith` then neither verifies either.

Its hard for me to know whether the above example is representative of the kinds of problems we are having.  But, the other problematic cases are more complicated and we need to simplify them further, etc.

### What type of operating system are you experiencing the problem on?

Linux

**Comment by DavePearce** [comment]
I should add that part of the issue here appears to be a greater instability of verification.  For example, using `/randomSeed:XX` for values of `XX` in `0 .. 10` gives the following:

```
[XX=0] FAIL
[XX=1] SUCCESS
[XX=2] SUCCESS
[XX=3] SUCCESS
**[XX=4] FAIL**
[XX=5] SUCCESS
**[XX=6] FAIL**
**[XX=7] FAIL**
[XX=8] SUCCESS
[XX=9] SUCCESS
[XX=10] SUCCESS
```
In contrast, with Dafny version `3.13.0` it verifies every single time.

**Comment by atomb** [comment]
Yes, I'm almost certain that it has to do with the Z3 version, since the other differences between 3.13 and 4.0 don't seem relevant to this example.

**Comment by atomb** [comment]
Another way to get this to work with Dafny 4.0 is the following arcane incantation:

```
$ dafny verify --boogie /proverOpt:O:smt.arith.solver=6 issue-3758.dfy

Dafny program verifier finished with 4 verified, 0 errors
```

For a little more color, 6 is the default value for that Z3 parameter in Z3 4.12.1, and it generally works best on most SMT benchmarks out there, especially those that do anything nonlinear. However, there are quite a few existing Dafny programs that break with that value. Dafny instead sets the value to 2 (and has for many years). I investigated the difference between 2 and 6 before the Dafny 4.0 release, and decided that changing the Dafny default to 6 would break more programs than it fixes. However, the current status isn't great. There isn't one specific value for that parameter that works uniformly well. I do suspect there's a possibility to make some changes in the Dafny -> SMT encoding that'd help, but those aren't available yet.

I'm increasingly wondering whether a Dafny-specific flag that's less inscrutable than the one above would be worth providing. I'd want to make sure it's available on a per-module level with `{:options ...}`, too.

**Comment by DavePearce** [comment]
Hey all,

So I can confirm this at least:

```
> dafny /proverOpt:PROVER_PATH=$HOME/pkg/dafny-4.0/z3/bin/z3-4.8.5 quadcost.dfy

Dafny program verifier finished with 4 verified, 0 errors
```

versus

```
 dafny /proverOpt:PROVER_PA dafny /proverOpt:PROVER_PATH=$HOME/pkg/dafny-4.0/z3/bin/z3-4.12.1 quadcost.dfy
/home/djp/projects/dafny/quadcost.dfy(16,25): Error: the calculation step between the previous line and this line might not hold

Dafny program verifier finished with 3 verified, 1 errorTH=$HOME/pkg/dafny-4.0/z3/bin/z3-4.12.1 quadcost.dfy
/home/djp/projects/dafny/quadcost.dfy(16,25): Error: the calculation step between the previous line and this line might not hold

Dafny program verifier finished with 3 verified, 1 error
```

I can have a play around with the `smt.arith.solver` option.   I was assuming that is the same as the `/arith` option?

**Comment by muchang** [comment]
Here are some other cases that are also related to the different z3 versions. https://github.com/dafny-lang/dafny/issues/3690.

**Comment by atomb** [comment]
Confusingly, the `smt.arith.solver` parameter to Z3 and `/arith` parameter to Dafny are almost entirely unrelated. My hope is that in the long run Dafny users won't have to worry about either, but there are unfortunately still cases where it is necessary to change one or the other.

**Comment by DavePearce** [comment]
> Confusingly, the smt.arith.solver parameter to Z3 and /arith parameter to Dafny are almost entirely unrelated

Ok, so that's good to know as I wasn't aware.  So, what does Dafny's `/arith` parameter do then?

**Comment by atomb** [comment]
> > Confusingly, the smt.arith.solver parameter to Z3 and /arith parameter to Dafny are almost entirely unrelated
> 
> Ok, so that's good to know as I wasn't aware. So, what does Dafny's `/arith` parameter do then?

It controls whether Dafny introduces wrappers around arithmetic operations, usable in triggers. For higher values of its parameter, it makes those wrappers uninterpreted functions and introduces axioms about them that can be used in place of Z3's arithmetic decision procedure(s).
